src/core/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,298 @@
+      setCurrentIndex: (index: number | ((old: number) => void)) => void;
+      slideCount: number;
+      setSlideCount: (length: number | ((old: number) => void)) => void;
+      keysPressed: Record<string, boolean>;
+    }
+  | undefined
+>(undefined);
+
+export const Carousel = (props: CarouselProps) => {
+  const { currentIndex: userCurrentIndex = 0, children, ...rest } = props;
+
+  useTheme();
+
+  const [currentIndex, setCurrentIndex] = React.useState(userCurrentIndex);
+  const [slideCount, setSlideCount] = React.useState(0);
+
+  const [keysPressed, setKeysPressed] = React.useState<Record<string, boolean>>(
+    {},
+  );
+
+  const handleKeyDown = (event: React.KeyboardEvent) => {
+    if (event.altKey || event.ctrlKey || event.metaKey) {
+      return;
+    }
+
+    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
+      setKeysPressed((old) => ({ ...old, [event.key]: true }));
+    }
+
+    switch (event.key) {
+      case 'ArrowLeft': {
+        setCurrentIndex((old) => (slideCount + old - 1) % slideCount);
+        break;
+      }
+      case 'ArrowRight': {
+        setCurrentIndex((old) => (slideCount + old + 1) % slideCount);
+        break;
+      }
+    }
+  };
+
+  const handleKeyUp = (event: React.KeyboardEvent) => {
+    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
+      setKeysPressed((old) => ({ ...old, [event.key]: false }));
+    }
+  };
+
+  return (
+    <div
+      className='iui-carousel'
+      aria-roledescription='carousel'
+      tabIndex={0}
+      onKeyDown={handleKeyDown}
+      onKeyUp={handleKeyUp}
+      {...rest}
+    >
+      <CarouselContext.Provider
+        value={{
+          currentIndex,
+          setCurrentIndex,
+          slideCount,
+          setSlideCount,
+          keysPressed,
+        }}
+      >
+        {children}
+      </CarouselContext.Provider>
+    </div>
+  );
+};
+
+const CarouselSlider = (props: React.ComponentPropsWithoutRef<'ol'>) => {
+  const context = React.useContext(CarouselContext);
+
+  if (!context) {
+    throw new Error('CarouselSlider must be used within Carousel');
+  }
+
+  const { currentIndex, setSlideCount } = context;
+
+  React.useLayoutEffect(() => {
+    setSlideCount(React.Children.count(props.children));
+  }, [props.children, setSlideCount]);
+
+  const items = React.useMemo(
+    () =>
+      React.Children.map(props.children, (child, index) =>
+        React.isValidElement(child)
+          ? React.cloneElement(child, { 'aria-hidden': index !== currentIndex })
+          : child,
+      ) ?? [],
+    [props.children, currentIndex],
+  );
+
+  return (
+    <ol
+      {...props}
+      className={cx('iui-carousel-slider', props.className)}
+      style={{
+        // height: '100%',
+        transform: `translateX(-${currentIndex * 100}%)`,
+        ...props.style,
+      }}
+    >
+      {items}
+    </ol>
+  );
+};
+
+const CarouselSlide = (props: React.ComponentPropsWithoutRef<'li'>) => {
```

Comment:
Have you considered passing elements as prop straight to CarouselSlider as array of `ReactNode` and then map those straight to `li` elements in there? Cause this component by itself does nothing

Reply:
I think it's ok for this component to not "do anything". It just represents a DOM element, which is expected in the composition approach.

I have considered accepting an array of slides in the parent though, as an alternative to the whole composition approach. But it requires too many props if we want it to be controlled.

```tsx
<Carousel slides={[<div>, <div>, <div>]} current={current} onSlideChange={() => {}} />
```